lib/pbench/test/unit/agent/test_tool_data_sink.py

```diff
@@ -284,13 +288,194 @@ def mock_make_server(host, port, app, *args, **kwargs):
-        with stop() and wait() methods.
+        with stop() and wait() methods from separate threads.
+
+        There are a number of scenarios for the order of operations between
+        threads that we need to test.  We list them here using "MainThr" as the
+        name of the "main thread" which _creates_ the WSGI thread, and "WsgiThr"
+        as the name of the created WSGI thread invoking the "run" method.
+
+        References:
+            .wait() called at
+                line 184 (in .stop() method)
+                line 833 (in __enter__ method)
+            .stop() called at
+                line 886 (in __exit__ method)
+
+        Scenario A:
+
+          * MainThr creates WSGI thread (WsgiThr not running)
+          * MainThr calls stop()
+          * WsgiThr starts running
+          * WsgiThr reports err_code == 0
+
+        Scenario B:
+
+          * MainThr creates WSGI thread
+          * WsgiThr starts running
+          * WsgiThr reports err_code == 0
+          * MainThr calls stop()
+
+        Scenario C:
+
+          * MainThr creates WSGI thread (WsgiThr not running)
+          * MainThr calls stop()
+          * WsgiThr starts running
+          * WsgiThr reports err_code > 0
+
+        Scenario D:
+
+          * MainThr creates WSGI thread
+          * WsgiThr starts running
+          * WsgiThr reports err_code > 0
+          * MainThr calls stop()
+
+        Scenario E:
+
+          * MainThr creates WSGI thread (WsgiThr not running)
+          * MainThr calls stop()
+          * WsgiThr starts running
+          * WsgiThr reports err_code < 0
+
+        Scenario F:
+
+          * MainThr creates WSGI thread
+          * WsgiThr starts running
+          * WsgiThr reports err_code < 0
+          * MainThr calls stop()
         """
-        assert len(caplog.records) == 0
+
+        def wsgi_run(scenario, wsgi_server, trace):
+            ret_val = None
+            if scenario in ("A", "C", "E"):
+                time.sleep(0.1)
+            try:
+                trace.append("WsgiThr - run")
+                wsgi_server.run(_test_app)
+            except Exception as exc:
+                ret_val = exc
+            return ret_val
+
+        def do_wait(scenario, wsgi_server, trace):
+            if scenario in ("B", "D", "F"):
+                time.sleep(0.1)
+            trace.append("MainThr - wait")
+            err_text, err_code = wsgi_server.wait()
+            return err_text, err_code
+
+        def do_stop(scenario, wsgi_server, trace):
+            if scenario in ("B", "D", "F"):
+                time.sleep(0.1)
+            trace.append("MainThr - stop")
+            wsgi_server.stop()
+
+        hostnames = dict(
+            A="host.example.com",
+            B="host.example.com",
+            C="oserror.example.com",
+            D="oserror.example.com",
+            E="exception.example.com",
+            F="exception.example.com",
+        )
```

Comment:
It's kinda cute that you're using the hostname to send messages to the WSGI server, but I think a comment here documenting how the messages are encoded ("starts with `exception`") and what they mean ("raise at this operation") would eliminate some potential future confusion.

Reply:
I had no idea that `example.com` is a [reserved domain](https://www.iana.org/domains/reserved)!  :-)

(Top-level domains `.test` and `.invalid` are [also available](https://tools.ietf.org/html/rfc2606) for these uses.)

Reply:
Fixed.